Test failure in test_update_resized_header in io.fits on Windows

If I install astropy:

```
pip install -e .
```

then go to a different directory and run the ``test_update_resized_header`` test, I get the following failure:

```
> pytest -p no:warnings --astropy-header --pyargs astropy -k update_resized
================================================= test session starts =================================================
platform win32 -- Python 3.7.4, pytest-5.3.4, py-1.8.1, pluggy-0.13.1

Running tests with Astropy version 4.2.dev136+g399f7e11f.
Running tests in astropy.

Date: 2020-05-28T16:23:44

Platform: Windows-10-10.0.18362-SP0

Executable: c:\programdata\miniconda3\python.exe

Full Python Version:
3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]

encodings: sys: utf-8, locale: cp1252, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions:
Numpy: 1.17.4
Scipy: 1.3.3
Matplotlib: 3.1.2
h5py: 2.10.0
Pandas: 0.25.3

Using Astropy options: remote_data: none.

rootdir: C:\Users\Thomas Robitaille
plugins: hypothesis-5.3.0, arraydiff-0.3, astropy-header-0.1.2, cov-2.8.1, doctestplus-0.5.0, filter-subpackage-0.1.1, openfiles-0.4.0, remotedata-0.3.2
collected 14725 items / 14723 deselected / 17 skipped

projects\astropy\astropy\io\fits\tests\test_hdulist.py F.                                                        [100%]

====================================================== FAILURES =======================================================
___________________________________ TestHDUListFunctions.test_update_resized_header ___________________________________

self = <astropy.io.fits.tests.test_hdulist.TestHDUListFunctions object at 0x000002544110F948>

    @pytest.mark.filterwarnings("ignore:Assigning the 'data' attribute is an "
                                "inherently unsafe operation")
    def test_update_resized_header(self):
        """
        Test saving updates to a file where the header is one block smaller
        than before, and in the case where the heade ris one block larger than
        before.
        """

        data = np.arange(100)
        hdu = fits.PrimaryHDU(data=data)
        idx = 1
        while len(str(hdu.header)) <= 2880:
            hdu.header[f'TEST{idx}'] = idx
            idx += 1
        orig_header = hdu.header.copy()
        hdu.writeto(self.temp('temp.fits'))

        with fits.open(self.temp('temp.fits'), mode='update') as hdul:
            while len(str(hdul[0].header)) > 2880:
                del hdul[0].header[-1]

        with fits.open(self.temp('temp.fits')) as hdul:
            assert hdul[0].header == orig_header[:-1]
            assert (hdul[0].data == data).all()

        with fits.open(self.temp('temp.fits'), mode='update') as hdul:
            idx = 101
            while len(str(hdul[0].header)) <= 2880 * 2:
                hdul[0].header[f'TEST{idx}'] = idx
                idx += 1
            # Touch something in the data too so that it has to be rewritten
>           hdul[0].data[0] = 27

projects\astropy\astropy\io\fits\tests\test_hdulist.py:657:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
projects\astropy\astropy\io\fits\hdu\hdulist.py:387: in __exit__
    self.close(output_verify=output_verify)
projects\astropy\astropy\io\fits\hdu\hdulist.py:970: in close
    self.flush(output_verify=output_verify, verbose=verbose)
projects\astropy\astropy\io\fits\util.py:241: in wrapped
    func(*args, **kwargs)
projects\astropy\astropy\io\fits\hdu\hdulist.py:849: in flush
    self._flush_update()
projects\astropy\astropy\io\fits\hdu\hdulist.py:1264: in _flush_update
    self._flush_resize()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = [<astropy.io.fits.hdu.image.PrimaryHDU object at 0x00000254411178C8>]

    def _flush_resize(self):
        """
        Implements flushing changes in update mode when parts of one or more HDU
        need to be resized.
        """

        old_name = self._file.name
        old_memmap = self._file.memmap
        name = _tmp_name(old_name)

        if not self._file.file_like:
            old_mode = os.stat(old_name).st_mode
            # The underlying file is an actual file object.  The HDUList is
            # resized, so we need to write it to a tmp file, delete the
            # original file, and rename the tmp file to the original file.
            if self._file.compression == 'gzip':
                new_file = gzip.GzipFile(name, mode='ab+')
            elif self._file.compression == 'bzip2':
                if not HAS_BZ2:
                    raise ModuleNotFoundError(
                        "This Python installation does not provide the bz2 module.")
                new_file = bz2.BZ2File(name, mode='w')
            else:
                new_file = name

            with self.fromfile(new_file, mode='append') as hdulist:

                for hdu in self:
                    hdu._writeto(hdulist._file, inplace=True, copy=True)
                if sys.platform.startswith('win'):
                    # Collect a list of open mmaps to the data; this well be
                    # used later.  See below.
                    mmaps = [(idx, _get_array_mmap(hdu.data), hdu.data)
                             for idx, hdu in enumerate(self) if hdu._has_data]

                hdulist._file.close()
                self._file.close()
            if sys.platform.startswith('win'):
                # Close all open mmaps to the data.  This is only necessary on
                # Windows, which will not allow a file to be renamed or deleted
                # until all handles to that file have been closed.
                for idx, mmap, arr in mmaps:
                    if mmap is not None:
                        mmap.close()

            os.remove(self._file.name)

            # reopen the renamed new file with "update" mode
            os.rename(name, old_name)
            os.chmod(old_name, old_mode)

            if isinstance(new_file, gzip.GzipFile):
                old_file = gzip.GzipFile(old_name, mode='rb+')
            else:
                old_file = old_name

            ffo = _File(old_file, mode='update', memmap=old_memmap)

            self._file = ffo

            for hdu in self:
                # Need to update the _file attribute and close any open mmaps
                # on each HDU
                if hdu._has_data and _get_array_mmap(hdu.data) is not None:
                    del hdu.data
                hdu._file = ffo

            if sys.platform.startswith('win'):
                # On Windows, all the original data mmaps were closed above.
                # However, it's possible that the user still has references to
                # the old data which would no longer work (possibly even cause
                # a segfault if they try to access it).  This replaces the
                # buffers used by the original arrays with the buffers of mmap
                # arrays created from the new file.  This seems to work, but
                # it's a flaming hack and carries no guarantees that it won't
                # lead to odd behavior in practice.  Better to just not keep
                # references to data from files that had to be resized upon
                # flushing (on Windows--again, this is no problem on Linux).
                for idx, mmap, arr in mmaps:
                    if mmap is not None:
>                       arr.data = self[idx].data.data
E                       DeprecationWarning: Assigning the 'data' attribute is an inherently unsafe operation and will be removed in the future.

projects\astropy\astropy\io\fits\hdu\hdulist.py:1357: DeprecationWarning
============================= 1 failed, 1 passed, 17 skipped, 14723 deselected in 11.05s ==============================
```

Somehow the warning capture (see test decorator) is not working. Same goes for ``test_io_manipulation``. Note that this only fails if I'm in a folder where ``setup.cfg`` has no effect. I'll open a separate issue, but we should make sure that ``pytest --pyargs astropy`` always works even when no ``setup.cfg`` is present.

--- a/astropy/io/fits/hdu/hdulist.py
+++ b/astropy/io/fits/hdu/hdulist.py
@@ -1,6 +1,5 @@
 # Licensed under a 3-clause BSD style license - see PYFITS.rst
 
-
 import gzip
 import itertools
 import os
@@ -354,7 +353,7 @@ class HDUList(list, _Verify):
             self._try_while_unread_hdus(super().__setitem__, _key, hdu)
         except IndexError:
             raise IndexError('Extension {} is out of bound or not found.'
-                            .format(key))
+                             .format(key))
 
         self._resize = True
         self._truncate = False
@@ -803,7 +802,7 @@ class HDUList(list, _Verify):
 
         if self._file.mode not in ('append', 'update', 'ostream'):
             warnings.warn("Flush for '{}' mode is not supported."
-                         .format(self._file.mode), AstropyUserWarning)
+                          .format(self._file.mode), AstropyUserWarning)
             return
 
         save_backup = self._open_kwargs.get('save_backup', False)
@@ -1354,7 +1353,10 @@ class HDUList(list, _Verify):
                 # flushing (on Windows--again, this is no problem on Linux).
                 for idx, mmap, arr in mmaps:
                     if mmap is not None:
-                        arr.data = self[idx].data.data
+                        # https://github.com/numpy/numpy/issues/8628
+                        with warnings.catch_warnings():
+                            warnings.simplefilter('ignore', category=DeprecationWarning)
+                            arr.data = self[idx].data.data
                 del mmaps  # Just to be sure
 
         else:

--- a/astropy/io/fits/tests/test_hdulist.py
+++ b/astropy/io/fits/tests/test_hdulist.py
@@ -619,11 +619,6 @@ class TestHDUListFunctions(FitsTestCase):
         with fits.open(self.temp('temp.fits')) as hdul:
             assert (hdul[0].data == data).all()
 
-    # This test used to fail on Windows - if it fails again in future, see
-    # https://github.com/astropy/astropy/issues/5797
-    # The warning appears on Windows but cannot be explicitly caught.
-    @pytest.mark.filterwarnings("ignore:Assigning the 'data' attribute is an "
-                                "inherently unsafe operation")
     def test_update_resized_header(self):
         """
         Test saving updates to a file where the header is one block smaller

--- a/astropy/io/fits/tests/test_image.py
+++ b/astropy/io/fits/tests/test_image.py
@@ -239,11 +239,6 @@ class TestImageFunctions(FitsTestCase):
         with fits.open(self.temp('test.fits')) as hdul:
             assert hdul[0].name == 'XPRIMARY2'
 
-    # This test used to fail on Windows - if it fails again in future, see
-    # https://github.com/astropy/astropy/issues/5797
-    # The warning appears on Windows but cannot be explicitly caught.
-    @pytest.mark.filterwarnings("ignore:Assigning the 'data' attribute is an "
-                                "inherently unsafe operation")
     def test_io_manipulation(self):
         # Get a keyword value.  An extension can be referred by name or by
         # number.  Both extension and keyword names are case insensitive.

